### PR TITLE
Container Instance: Add Assign MSI Identity for container create

### PR DESCRIPTION
**Pull request by samkreter**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py [hunk 1]
```diff
@@ -210,6 +217,21 @@ def create_container(cmd,
     return sdk_no_wait(no_wait, container_group_client.create_or_update, resource_group_name, name, cgroup)
 
 
+def _build_identities_info(identities):
+    identities = identities or []
```
Comment (adewaleo) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py]: Also instead of building the identity_types string using str.join and hardcoded strings, you should use the [ResourceIdentityType](https://github.com/Azure/azure-sdk-for-python/blob/master/azure-mgmt-containerinstance/azure/mgmt/containerinstance/models/container_instance_management_client_enums.py) enum. You can import it from azure.mgmt.containerinstance or use cmd.get_model().

P.S. Side note but there is a model called [ContainerGroupIdentity](https://github.com/Azure/azure-sdk-for-python/blob/master/azure-mgmt-containerinstance/azure/mgmt/containerinstance/models/container_group_identity.py) that can be used to build the identity object you pass to [ContainerGroup](https://github.com/Azure/azure-sdk-for-python/blob/master/azure-mgmt-containerinstance/azure/mgmt/containerinstance/models/container_group.py). If this model makes your method more concise, consider using it.

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/tests/latest/test_container_commands.py [hunk 2]
```diff
@@ -125,7 +127,7 @@ def test_container_create_with_acr(self, resource_group, resource_group_location
         registry_username = 'clitestregistry1'
         registry_server = '{}.azurecr.io'.format(registry_username)
         image = '{}/nginx:latest'.format(registry_server)
-        password = 'EsYA8DyuSaNv+3HyGyHoiIIafktf0kE6'
+        password = 'rylgTHkXbBp6k/F0Gd3lOpVnGU0dVEq4'
```
Comment (adewaleo) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/tests/latest/test_container_commands.py]: Out of curiosity, why did you change the password?

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/tests/latest/test_container_commands.py [hunk 3]
```diff
@@ -55,7 +55,8 @@ def test_container_create(self, resource_group, resource_group_location):
         self.cmd('container create -g {rg} -n {container_group_name} --image {image} --os-type {os_type} '
                  '--ip-address {ip_address_type} --dns-name-label {dns_name_label} --ports {ports} --cpu {cpu} --memory {memory} '
                  '--command-line {command} -e {env} --restart-policy {restart_policy} '
-                 '--secrets {secrets} --secrets-mount-path {secrets_mount_path}',
+                 '--secrets {secrets} --secrets-mount-path {secrets_mount_path} '
+                 '--assign --assign-identity',
```
Comment (adewaleo) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/tests/latest/test_container_commands.py]: Is it possible to add a test for user assigned identities, or at least to unit test your user assigned identities method?

Comment (samkreter) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/tests/latest/test_container_commands.py]: That would require creating an actual user assigned identity to run when creating the recordings. Once recorded everything would work fine but I didn't know if it was worth adding the extra dependency for recording the tests. 

Do you think it's something that is worth it?

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/tests/latest/test_container_commands.py (hunk 2 above)
Comment (adewaleo) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/tests/latest/test_container_commands.py]: Okay, I just read the comment. But please bear with me. I don't understand conceptually why the password has to be automatically re-generated every time the test is run live. Does this have to do with the fact that the test is created using the clitestregistry1 nginx image?

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/tests/latest/test_container_commands.py (hunk 3 above)
Comment (adewaleo) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/tests/latest/test_container_commands.py]: If it would not be too difficult or add too much overhead it would definitely be worth it. That way we can catch any regressions that occur if anyone modifies the method you add. If we do not add this test to the scenario test, we should at least add some unit tests to the method you added for building the identity list.

Comment (adewaleo) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/tests/latest/test_container_commands.py]: [This](https://github.com/Azure/azure-cli/blob/be891a9360988aacfbc1df1f6853f5d7dc8e3178/src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py#L2212-L2224) would be helpful for creating a temporary identity. You can access the identity id by referencing the id field in the json object, which can be retrieved from the execution result by calling get_output_in_json().

Comment (samkreter) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/tests/latest/test_container_commands.py]: That's a good solution, Adding it now.

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py [hunk 4]
```diff
@@ -210,6 +217,22 @@ def create_container(cmd,
...
+    identity_type = ResourceIdentityType.none
+    if not identities or MSI_LOCAL_ID in identities:
+        identity_type = ResourceIdentityType.system_assigned
+    external_identities = [x for x in identities if x != MSI_LOCAL_ID]
+    if external_identities and identity_type == ResourceIdentityType.system_assigned:
+        identity_type = ResourceIdentityType.system_assigned_user_assigned
+    elif external_identities:
+        identity_type = ResourceIdentityType.user_assigned
+    identity = ContainerGroupIdentity(type=identity_type)
+    if external_identities:
```
Comment (adewaleo) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py]: Why not construct the identity object after checking if there are external_identities?

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/tests/latest/test_container_commands.py [hunk 5]
```diff
@@ -115,6 +115,69 @@ def test_container_create(self, resource_group, resource_group_location):
                              '[0].containers[0].resources.requests.cpu', cpu),
                          self.check('[0].containers[0].resources.requests.memoryInGb', memory)])
 
+    # Test create container using managed identities.
```
Comment (adewaleo) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/tests/latest/test_container_commands.py]: Thanks for adding this test!

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py (hunk 4 above)
Comment (adewaleo) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py]: @samkreter let me know your thoughts on this.

Comment (samkreter) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py]: That would add an extra step of either having an if else creating the object with or without the user assigned identities or reassigning the external_identities variable to be in the correct format. This way seemed to me to be the most readable.

Comment (samkreter) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py]: That would add an extra step of either having an if else creating the object with or without the user assigned identities or reassigning the external_identities variable to be in the correct format. This way seemed to me to be the most readable.
